include/chainbase/chainrocks.hpp

```diff
@@ -0,0 +1,822 @@
+            _state[modified_value.first] = std::move(modified_value.second);
+         }
+      }
+
+      /// Effectively reintroduce any removed key/value pairs from
+      /// `_state` back into `_state`.
+      void _undo_removed_values(undo_state&& head) {
+         for (auto&& removed_value : head._removed_values) {
+            _state[removed_value.first] = std::move(removed_value.second);
+         }
+      }
+
+      // Areg's review:
+      // "In Chainbase, each new object had a unique ID. So it wasn't
+      // possible to have a removed value in head_minus_one and a new
+      // value in head that would conflict. But with this key value
+      // system, it is possible to have a removed value entry for key
+      // K in head_minus_one and a new value entry for the same key K
+      // in head. In this case, squash needs to ensure that the new
+      // head has neither a removed value entry nor a new value entry
+      // but rather a modified value entry for the key K."
+
+      // head-minus-one REMOVE
+      // head           NEW
+      // head-minus-one MODIFY
+
+      /// Effectively squash `_new_keys` of the previous two
+      /// `undo_state` objects on the `_stack` together.
+      void _squash_new_keys(undo_state&& head, undo_state& head_minus_one) {
+         for (auto&& key : head._new_keys) {
+            head_minus_one._new_keys.insert(std::move(key));
+         }
+      }
+
+      /// Effectively squash `_modifed_values` of the previous two
+      /// `undo_state` objects on the `_stack` together.
+      void _squash_modified_values(undo_state&& head, undo_state& head_minus_one) {
+         for (auto&& value : head._modified_values) {
+            if (head_minus_one._new_keys.find(value.first) != head_minus_one._new_keys.cend()) {
+               continue;
+            }
+            if (head_minus_one._modified_values.find(value.first) != head_minus_one._modified_values.cend()) {
+               continue;
+            }
+            assert(head_minus_one._removed_values.find(value.first) == head_minus_one._removed_values.cend());
+            head_minus_one._modified_values[value.first] = std::move(value.second);
+         }
+      }
+
+      /// Effectively squash `_removed_values` of the previous two
+      /// `undo_state` objects on the `_stack` together.
+      void _squash_removed_values(undo_state&& head, undo_state& head_minus_one) {
+         for (auto&& value : head._removed_values) {
+            if (head_minus_one._new_keys.find(value.first) != head_minus_one._new_keys.cend()) {
+               head_minus_one._new_keys.erase(std::move(value.first));
+               continue;
+            }
+
+            auto iter{head_minus_one._modified_values.find(value.first)};
+            if (iter != head_minus_one._modified_values.cend()) {
+               head_minus_one._removed_values[iter->first] = std::move(iter->second);
+               head_minus_one._modified_values.erase(std::move(value.first));
```

Comment:
`head_minus_one._modified_values.erase(iter);` would be even better.

Reply:
Resolved in [28f37d5c88015e37e687a4d79b33de440345e377](https://github.com/EOSIO/chainbase/blob/28f37d5c88015e37e687a4d79b33de440345e377/include/chainbase/chainrocks.hpp#L576-L594).